keyboards/wuque/mammoth75x/mammoth75x.c

```diff
@@ -0,0 +1,75 @@
+/* Copyright 2021 wuquestudio
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "mammoth75x.h"
+
+#ifdef VIA_ENABLE
+static uint8_t encoder_state[ENCODERS] = {0};
+static keypos_t encoder_cw[ENCODERS] = ENCODERS_CW_KEY;
+static keypos_t encoder_ccw[ENCODERS] = ENCODERS_CCW_KEY;
+
+void encoder_action_unregister(void)
+{
+    for (int index = 0; index < ENCODERS; ++index)
+    {
+        if (encoder_state[index])
+        {
+            keyevent_t encoder_event = (keyevent_t){
+                .key = encoder_state[index] >> 1 ? encoder_cw[index] : encoder_ccw[index],
+                .pressed = false,
+                .time = (timer_read() | 1)};
+            encoder_state[index] = 0;
+            action_exec(encoder_event);
+        }
+    }
+}
+void encoder_action_register(uint8_t index, bool clockwise)
+{
+    keyevent_t encoder_event = (keyevent_t){
+        .key = clockwise ? encoder_cw[index] : encoder_ccw[index],
+        .pressed = true,
+        .time = (timer_read() | 1)};
+    encoder_state[index] = (clockwise ^ 1) | (clockwise << 1);
+    action_exec(encoder_event);
+}
+
+void matrix_scan_kb(void)
+{
+    encoder_action_unregister();
+    matrix_scan_user();
+}
+
+bool encoder_update_kb(uint8_t index, bool clockwise)
+{
+    encoder_action_register(index, clockwise);
+    return true;
+};
+#else
+bool encoder_update_kb(uint8_t index, bool clockwise) {
+    return encoder_update_user(index, clockwise);
+}
+
+bool encoder_update_user(uint8_t index, bool clockwise) {
+    if (index == 0) { /* First encoder */
+        if (clockwise) {
+            tap_code(KC_PGDN);
+        } else {
+            tap_code(KC_PGUP);
+        }
+    }
+    return true;
+}
```

Comment:
The user function would need to be set as weak, or if a user tries to change it, it will error out when compiling. 

Alternatively, I do recommend this: 
```suggestion
bool encoder_update_kb(uint8_t index, bool clockwise) {
    if (!encoder_update_user(index, clockwise)) { return false; }
    if (index == 0) { /* First encoder */
        if (clockwise) {
            tap_code(KC_PGDN);
        } else {
            tap_code(KC_PGUP);
        }
    }
    return true;
}
```

This basically checks the user function, and if that returns false, it will prevent the rest from running.  This gives a default behavior, but doesn't require you to add the user function (as it's already set, and set weakly in core files).

Reply:
ok, I think I get it!

Reply:
I will not make this mistake next time.

Reply:
Thank you for your time!

Reply:
No worries, it's a fairly recent change in implementation, to be honest.  But it also provides a better user experience. IMO.